cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go

```diff
@@ -412,13 +407,167 @@ func deploymentNameForEnv(envName string, clock clock.Clock) string {
 	return name[len(name)-cArmDeploymentNameLengthMax:]
 }
 
+// deploymentState returns the latests deployment if it is the same as the deployment within deploymentData or an error
+// otherwise.
+func (p *BicepProvider) deploymentState(
+	ctx context.Context,
+	deploymentData *deploymentDetails,
+	currentParamsHash string) (*armresources.DeploymentExtended, error) {
+
+	p.console.ShowSpinner(ctx, "Comparing deployment state", input.Step)
+	prevDeploymentResult, err := p.latestDeploymentResult(ctx, deploymentData.Target)
+	if err != nil {
+		return nil, fmt.Errorf("deployment state error: %w", err)
+	}
+
+	var templateHash string
+	createHashResult, err := p.deploymentsService.CalculateTemplateHash(
+		ctx, p.env.GetSubscriptionId(), deploymentData.CompiledBicep.RawArmTemplate)
+	if err != nil {
+		return nil, fmt.Errorf("can't get hash from current template: %w", err)
+	}
+
+	if createHashResult.TemplateHash != nil {
+		templateHash = *createHashResult.TemplateHash
+	}
+
+	if !prevDeploymentEqualToCurrent(ctx, prevDeploymentResult, templateHash, currentParamsHash) {
+		return nil, fmt.Errorf("deployment state has changed")
+	}
+
+	return prevDeploymentResult, nil
+}
+
+// prevDeploymentResult looks and finds a previous deployment for the current azd project.
+func (p *BicepProvider) latestDeploymentResult(
+	ctx context.Context,
+	scope infra.Scope,
+) (*armresources.DeploymentExtended, error) {
+	deployments, err := p.findCompletedDeployments(ctx, p.env.GetEnvName(), scope, "")
+	// findCompletedDeployments returns error if no deployments are found
+	// No need to check for empty list
+	if err != nil {
+		return nil, err
+	}
+
+	if len(deployments) > 1 {
+		// If more than one deployment found, ignore the prev-deployment
+		return nil, fmt.Errorf("more than one previous deployment match.")
+	}
+
+	return deployments[0], nil
+}
+
+// parametersHash generates a hash from each parameter as [name-type-value]
+func parametersHash(templateParameters azure.ArmTemplateParameterDefinitions, params azure.ArmParameters) (string, error) {
+	hash256 := sha256.New()
+
+	// the order of the parameters is important for creating a hash out of them
+	var orderedParamKeys []string
```

Comment:
Sort of imagined that we might just use `json.Marshal` on the `params` value here to get a `[]byte` and hash that.  `json.Marshal` is documented as sorting the keys of the map, so this should be a stable algorithm.  I have a slight preference to doing that instead of doing more of it by hand and using `gob`, as others would be able to replicate our work.

Was that something you thought about and discarded because it wouldn't work?

Reply:
Can't use it only for the `params` value.
As a customer, I can add any parameter name/value to `parameters.json` and it can or can be not a real parameter defined in bicep.

I have updated the code to create a map of the parameter definitions from the template with what it would be its final value (after applying the parameters file on them).  Then using json.marshall as you correctly pointed out.

I was also using the parameter type in the hash, but I just realized any change to the parameter definition would break the previous template hash. So here we can only matter about the param name and its value